src/rebuild/srv.c

```diff
@@ -652,25 +707,26 @@ rebuild_global_pool_tracker_create(struct ds_pool *pool, uint32_t ver,
 {
 	struct rebuild_global_pool_tracker *rgt;
 	unsigned int node_nr;
-	unsigned int array_size;
+	struct pool_domain *doms;
+	int i;
 	int rc = 0;
 
 	D_ALLOC_PTR(rgt);
 	if (rgt == NULL)
 		return -DER_NOMEM;
-	D_INIT_LIST_HEAD(&rgt->rgt_list);
 
-	node_nr = pool_map_node_nr(pool->sp_map);
-	array_size = roundup(node_nr, NBBY) / NBBY;
-	rgt->rgt_bits_size = node_nr;
+	D_INIT_LIST_HEAD(&rgt->rgt_list);
+	node_nr = pool_map_find_nodes(pool->sp_map, PO_COMP_ID_ALL, &doms);
+	if (node_nr < 0)
+		D_GOTO(out, rc = node_nr);
 
-	D_ALLOC_ARRAY(rgt->rgt_scan_bits, array_size);
-	if (rgt->rgt_scan_bits == NULL)
+	D_ALLOC_ARRAY(rgt->rgt_servers, node_nr);
+	if (rgt->rgt_servers == NULL)
 		D_GOTO(out, rc = -DER_NOMEM);
 
-	D_ALLOC_ARRAY(rgt->rgt_pull_bits, array_size);
-	if (rgt->rgt_pull_bits == NULL)
-		D_GOTO(out, rc = -DER_NOMEM);
+	for (i = 0; i < node_nr; i++)
+		rgt->rgt_servers[i].rank = doms[i].do_comp.co_rank;
```

Comment:
maybe not a problem in this patch, but seems there is some problems for the rank usage.
here set the rank as #rank in primary rank, and compare this rank in rebuild_leader_set_status(),
but some IV path used source #rank is logical rank within pool group, for example what set in rpt_create() -> crt_group_rank(pool->sp_group, &rank) -- that rank will be used as src_rank for rebuild_iv_update (for example in rebuild_tgt_status_check_ult).

Reply:
I think the rank inside the pool map should also the "logical" rank inside the pool group, instead of the rank in the primary group.

Anyway, I thought they are the same in our implementation?

Reply:
> I think the rank inside the pool map should also the "logical" rank inside the pool group, instead of the rank in the primary group.

No, rank inside pool map is primary rank.
> 
> Anyway, I thought they are the same in our implementation?

Possibly not same. For example a pool created with 4 ranks 0, 3, 4, 5, then each member will has a logical rank within that pool's sub-group of 0, 1, 2, 3. So they are different.

Reply:
> Anyway, I thought they are the same in our implementation?

You are right, there were some changes made them same (crt_group_secondary_modify()).

Reply:
True, but I think this is worth writing a comment in rebuild_leader_set_status, saying that we're converting an SG rank to the PG rank, without calling crt_group_rank_s2p, because update_pool_group in src/pool/srv_target.c makes sure any SG rank == PG rank.